### PR TITLE
GG-19644 Recreate caches after client node reconnected to cluster.

### PR DESCRIPTION
**Pull request by akuznetsov-gridgain**

### CODE REVIEW

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 1]
```diff
@@ -33,15 +41,30 @@
  */
 @Service
 public class TransactionManager {
+    /** */
+    private static final String ERR_CACHE_START = "Cannot start/stop cache within lock or transaction";
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Why we need this constant?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 2]
```diff
@@ -33,15 +41,30 @@
...
+    private int reconnectTimeout = 30_000;
+
     /**
      * @param ignite Ignite.
      */
     @Autowired
     protected TransactionManager(Ignite ignite) {
         this.ignite = ignite;
+
+        cachesStarters = new ConcurrentHashMap<>();
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: move to variable definition

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 3]
```diff
@@ -67,7 +90,117 @@ public Transaction txStart(TransactionConcurrency concurrency, TransactionIsolat
...
+    }
+
+    /**
+     * Execute action in transaction.
+     *
+     * @param name Action name.
+     * @param action Action to execute.
+     * @return Action result.
+     */
+    public <T> T doInTransaction(String name, Supplier<T> action) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Why you use Supplier when you code throw exception? Use Callable instead.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 1 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: inlined

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 2 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Moved

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 3 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: "Callable.call() throws Exception" and that a pain to handle it.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 4]
```diff
@@ -67,7 +90,118 @@ public Transaction txStart(TransactionConcurrency concurrency, TransactionIsolat
...
```
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Not really, just for loggin. May be make sense to remove "name".

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 5]
```diff
@@ -67,7 +90,118 @@ public Transaction txStart(TransactionConcurrency concurrency, TransactionIsolat
...
+
+            // TODO GG-19681: In future versions specific exception will be added.
+            boolean recreate = e instanceof IgniteException &&
+                msg != null &&
+                msg.startsWith(ERR_CACHE_START);
+
+            if (recreate) {
+                try {
+                    // Try to create caches, but client node may be disconnected.
+                    recreateCaches();
```
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Fixed

File: modules/web-console/frontend/app/modules/states/logout.state.js [hunk 6]
```diff
@@ -27,6 +27,7 @@ angular.module('ignite-console.states.logout', [
         controller: ['Auth', function(Auth) {Auth.logout();}],
         tfMetaTags: {
             title: 'Logout'
-        }
+        },
+        template: `<div class='splash'><div class='splash-wrapper'><div class='spinner'><div class='bounce1'></div><div class='bounce2'></div><div class='bounce3'></div></div><div class='splash-wellcome'>Logout...</div></div></div>`
```
Comment (nva) [inline modules/web-console/frontend/app/modules/states/logout.state.js]: Can be removed by appending logout().invalidateHttpSession(false) in org.apache.ignite.console.web.security.SecurityConfig#configure(org.springframework.security.config.annotation.web.builders.HttpSecurity)

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/errors/Errors.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        return e instanceof IgniteException &&
+            msg != null &&
+            msg.startsWith("Cannot start/stop cache within lock or transaction");
+    }
+
+    /**
+     * @param e Exception to check.
+     * @return Exception to throw.
+     */
+    public static RuntimeException checkDatabaseNotAvailable(RuntimeException e) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/errors/Errors.java]: May be convertException?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java [hunk 8]
```diff
@@ -44,30 +56,137 @@ protected TransactionManager(Ignite ignite) {
...
+
+        // For client nodes:
+        //  1. Try to execute action in transaction.
+        //  2. If failed, try to recreate caches.
+        //  3. If failed with IgniteClientDisconnectedException - await on reconnect future.
+        //  4. If all options failed, throw IgniteClientDisconnectedException that will be converted to 503 error.
+        try {
+            return doInTransaction0(action);
+        }
+        catch (Throwable e) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Why not RuntimeException?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/security/IgniteSessionRepository.java [hunk 9]
```diff
@@ -63,29 +74,51 @@ public IgniteSessionRepository setDefaultMaxInactiveInterval(int dfltMaxInactive
...
+            return ignite.getOrCreateCache(cfg);
+    }
+
     /** {@inheritDoc} */
     @Override public void save(ExpiringSession ses) {
-        cache.put(ses.getId(), new MapSession(ses));
+        try {
+            cache().put(ses.getId(), new MapSession(ses));
+        }
+        catch (IgniteClientDisconnectedException e) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/security/IgniteSessionRepository.java]: Why not RuntimeException?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java [hunk 10]
```diff
@@ -85,27 +99,29 @@ public Account getById(UUID accId) throws IllegalStateException {
...
+        return txMgr.doInTransaction("Find account by e-mail", () -> {
+            try {
+                Account account = accountsTbl.getByIndex(email);
+
+                if (account == null)
+                    throw new UsernameNotFoundException(email);
+
+                return account;
+            }
+            catch (IgniteException e) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java]: Why not RuntimeException?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java [hunk 11]
```diff
@@ -54,27 +56,39 @@
...
+        return txMgr.doInTransaction("Find account by ID", () -> {
+            try {
+                Account account = accountsTbl.load(accId);
+
+                if (account == null)
+                    throw new UsernameNotFoundException(accId.toString());
+
+                return account;
+            }
+            catch (IgniteException e) {
```
Comment (nva) [inline modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java]: Why not RuntimeException?

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/Utils.java [hunk 12]
```diff
@@ -125,4 +125,12 @@ public static String currentRequestOrigin() {
             .build()
             .toString();
     }
+
+    /**
+     * @param email E-mail.
+     * @return Normalized e-mail.
+     */
+    public static String normalizeEmail(String email) {
```
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/Utils.java]: Reverted

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/controller/RestExceptionHandler.java [hunk 13]
```diff
@@ -64,7 +67,31 @@
...
+    }
+
+    /**
+     * Handles client disconnected exceptions.
+     *
+     * @param ex Service exception.
+     * @param req Web request.
+     * @return {@link ErrorResponse} instance with error code and message.
+     */
+    @ExceptionHandler(value = {IgniteClientDisconnectedException.class})
```
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/controller/RestExceptionHandler.java]: Reverted.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java (hunk 11 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java]: Fixed.

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java (hunk 10 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/repositories/AccountsRepository.java]: Fixed

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/security/IgniteSessionRepository.java (hunk 9 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/security/IgniteSessionRepository.java]: Fixed

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 8 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: Fixed

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/errors/Errors.java (hunk 7 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/errors/Errors.java]: renamed

File: modules/web-console/frontend/app/modules/states/logout.state.js (hunk 6 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/frontend/app/modules/states/logout.state.js]: I will check

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: fixed

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java (hunk 3 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/tx/TransactionManager.java]: fixed

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/Utils.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (akuznetsov-gridgain) [reply modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/common/Utils.java]: deleted

File: modules/web-console/frontend/app/modules/states/logout.state.js (hunk 6 above)
Comment (akuznetsov-gridgain) [reply modules/web-console/frontend/app/modules/states/logout.state.js]: This setting has no effect.

**Pull request merged**